stratum/lib/p4runtime/sdn_controller_manager.cc

```diff
@@ -0,0 +1,445 @@
+    const absl::optional<std::string>& role_name,
+    const absl::optional<absl::uint128>& election_id) const {
+  absl::MutexLock l(&lock_);
+
+  if (!election_id.has_value()) {
+    return grpc::Status(grpc::StatusCode::PERMISSION_DENIED,
+                        "Request does not have an election ID.");
+  }
+
+  const auto& election_id_past_for_role =
+      election_id_past_by_role_.find(role_name);
+  if (election_id_past_for_role == election_id_past_by_role_.end()) {
+    return grpc::Status(grpc::StatusCode::PERMISSION_DENIED,
+                        "Only the primary connection can issue requests, but "
+                        "no primary connection has been established.");
+  }
+
+  if (election_id != election_id_past_for_role->second) {
+    return grpc::Status(grpc::StatusCode::PERMISSION_DENIED,
+                        "Only the primary connection can issue requests.");
+  }
+
+  return VerifyElectionIdIsActive(role_name, election_id, connections_);
+}
+
+grpc::Status SdnControllerManager::AllowRequest(
+    const p4::v1::WriteRequest& request) const {
+  absl::optional<std::string> role_name;
+#if P4RT_HAS_ROLES
+  if (!request.role().empty()) {
+    role_name = request.role();
+  }
+#endif
+
+  absl::optional<absl::uint128> election_id;
+  if (request.has_election_id()) {
+    election_id = absl::MakeUint128(request.election_id().high(),
+                                    request.election_id().low());
+  }
+  return AllowRequest(role_name, election_id);
+}
+
+grpc::Status SdnControllerManager::AllowRequest(
+    const p4::v1::SetForwardingPipelineConfigRequest& request) const {
+  absl::optional<std::string> role_name;
+#if P4RT_HAS_ROLES
+  if (!request.role().empty()) {
+    role_name = request.role();
+  }
+#endif
+
+  absl::optional<absl::uint128> election_id;
+  if (request.has_election_id()) {
+    election_id = absl::MakeUint128(request.election_id().high(),
+                                    request.election_id().low());
+  }
+  return AllowRequest(role_name, election_id);
+}
+
+int SdnControllerManager::ActiveConnections() const {
+  absl::MutexLock l(&lock_);
+  return connections_.size();
+}
+
+void SdnControllerManager::InformConnectionsAboutPrimaryChange(
+    const absl::optional<std::string>& role_name) {
+  VLOG(1) << "Informing all connections about primary connection change.";
+  for (const auto& connection : connections_) {
+    if (connection->GetRoleName() == role_name) {
+      SendArbitrationResponse(connection);
+    }
+  }
+}
+
+bool SdnControllerManager::PrimaryConnectionExists(
+    const absl::optional<std::string>& role_name) {
+  absl::optional<absl::uint128> election_id_past_for_role =
+      election_id_past_by_role_[role_name];
+
+  for (const auto& connection : connections_) {
+    if (connection->GetRoleName() == role_name &&
+        connection->GetElectionId() == election_id_past_for_role) {
+      return election_id_past_for_role.has_value();
```

Comment:
Is there a case that the election id for a role is an empty value? I believe that we always we always add or modify the SdnConnection with a valid election id (https://github.com/stratum/stratum/blob/82b2ff66659fc2324f213e3b3b10da081b4a704b/stratum/hal/lib/common/p4_service.cc#L583-L586)

Reply:
I think the spec allows backups without any election ID set:

The sender need not specify an election_id. If t

> he election_id is not specified, the sender's election_id is considered lower than any election_id, and the sender will thus never become primary.

Reply:
got it, maybe we can remove the check from the `p4_service`?

Reply:
I think I understood the logic here.


https://github.com/stratum/stratum/blob/4db73796d082973e63d3309b373343927ef19d4c/stratum/lib/p4runtime/sdn_controller_manager.h#L152-L154

Constructing the map with `absl::optional` as values allows liberal use of the index operator (`map[]`) without having to worry about inserting a key with a default value by accident. Which is what happens in a statement like this: `a = map[key];`. Looks like a read, but is actually a write into the map. With `absl::optional` values the key "presence check" becomes `map[key].has_value()` instead of `map.find()`.

Nonetheless, I simplified the check a bit by using the same patter as in other places.